flink/v1.17/flink/src/main/java/org/apache/iceberg/flink/sink/CachingTableSupplier.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.flink.sink;
+
+import java.time.Duration;
+import org.apache.flink.util.Preconditions;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.flink.TableLoader;
+import org.apache.iceberg.util.DateTimeUtil;
+import org.apache.iceberg.util.SerializableSupplier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A table loader that will only reload a table after a certain interval has passed. WARNING: This
+ * table loader should be used carefully when used with writer tasks. It could result in heavy load
+ * on a catalog for jobs with many writers.
+ */
+class CachingTableSupplier implements SerializableSupplier<Table> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CachingTableSupplier.class);
+
+  private final Table initialTable;
+  private final TableLoader tableLoader;
+  private final Duration tableRefreshInterval;
+  private long nextReloadTimeMs;
+  private transient Table table;
+
+  CachingTableSupplier(Table initialTable, TableLoader tableLoader, Duration tableRefreshInterval) {
+    Preconditions.checkArgument(initialTable != null, "initialTable cannot be null");
+    Preconditions.checkArgument(tableLoader != null, "tableLoader cannot be null");
+    Preconditions.checkArgument(
+        tableRefreshInterval != null, "tableRefreshInterval cannot be null");
+    this.initialTable = initialTable;
+    this.table = initialTable;
+    this.tableLoader = tableLoader;
+    this.tableRefreshInterval = tableRefreshInterval;
+    this.nextReloadTimeMs = System.currentTimeMillis() + tableRefreshInterval.toMillis();
+  }
+
+  @Override
+  public Table get() {
+    if (table == null) {
+      this.table = initialTable;
+    }
+    return table;
+  }
+
+  public void refresh() {
```

Comment:
That was how it was originally, @pvary noted it is better to explicitly refresh only during the checkpoint rather than at indeterminate times when `get()` is called.

Reply:
@stevenzwu: I am concerned about unintentionally refreshing data during operations where consistency is needed. If we do refresh behind the scenes, we can break the consistency requirement. We either have to comb through the code to fix all these issues, or we keep refresh a different method.

Reply:
got it. I am fine with that reasoning.

but with that, should `tableRefreshInterval(Duration)` take a simple boolean arg? Do we always want to tie the table refresh interval with checkpoint interval? do we Flink jobs checkpoint pretty frequently (like seconds)?

I am also thinking about in the future when we want to add coordinator broadcast and schema evolution support. it is always ok to refresh io along with schema/partition spec, right? I think yes. since we are refreshing table, those all come together, it is no harm to refresh them.

Reply:
The interval is a minimum threshold (it was named `minRefreshInterval` originally). We don't necessarily want to refresh at every checkpoint. Though we are doing that in the Kafka Connect sink and haven't had issues, but that probably is not being used at the same scale as Flink jobs.

It should be fine to refresh the schema, spec, and FileIO all together.

Reply:
once we use time threshold, then we won't refresh at checkpoint boundary anymore. those 2 are conflicting. we probably have to stay with only one (not both)

Reply:
The refresh is called in `prepareSnapshotPreBarrier()` so it should happen at a deterministic time. But It won't happen at every checkpoint.

Reply:
because of the time threshold check, some tasks may actually refresh handsome may skip (e.g. due to different init time)

Reply:
Yes that is possible. The tasks won't necessarily be in sync.

Reply:
I'm OK with refreshing every checkpoint, our catalog should be able to handle the load. Ryan early on had the thought of refreshing every n checkpoints which is another option.

Reply:
I am leaning toward purely time based refresh interval if there is no requirements for synchronizing the switch point at checkpoint barrier, which is always a bit tricky to guarantee. We might just need to enhance `IcebergFilesCommitter` to handle partition spec evolution. But that is for future scope of supporting schema and partition spec evolution in the sink.

Reply:
We'll wait for others to reply, from my perspective, either one works for FileIO refresh, but for schema evolution I think you'd want it on checkpoint boundaries.

Reply:
>  but for schema evolution I think you'd want it on checkpoint boundaries.

This is the assumption that I like us to challenge. It was also my initial assumption too. 

For partition spec evolution, we need to enhance `IcebergFilesCommitter` as a single manifest file is tied to one partition spec. if we have data files from two different partition specs (during transition period), committer can separate them out and commit in separate transactions.

For schema evolution, I am not sure if we need to do anything special if data files have 2 different schemas (during transition period).

In the future coordinator broadcast approach, it can be more complicated to apply the switch at checkpoint boundary and probably more difficult to guarantee it.  We can either add the complexity in the committer operator to enhance the handling of partition spec evolution or add the complexity in the coordinator broadcast implementation. I am wondering if the former is a bit cleaner and easier.

Reply:
We may want to ensure the schema, partition spec, and Flink schema are in sync between the appender factory and writer when schema evolution is introduced? That isn't an issue for now as those use the initial table only.

Reply:
> We may want to ensure the schema, partition spec, and Flink schema are in sync between the appender factory and writer when schema evolution is introduced?

Once every writer task refreshed the table object for schema or partition spec change, the latest table object can be used directly by appender factory from the `Supplier<Table>`.  do we need to align the cutover across all writer tasks at checkpoint boundary? 

Why can't every writer task refresh and apply schema and partition spec change independently? I see the need for committer to handle multiple partition specs, which needs to be enhanced. what else?

 >  schema evolution

I see two forms of automatic schema evolution. Maybe a little diverted from the scope of this PR.

1. Iceberg table schema is automatically updated out-of-band (e.g. by control plane). but we want to avoid the need to redeploy the Flink job to pick up the latest schema in the Iceberg table. This is the simpler form and easier to support. In this form, typically Iceberg table schema is updated way before data started to flow with new schema.

2. Incoming data changed (e.g. a new column is added). Flink writer detected that the Iceberg table schema doesn't have the new column, update the Iceberg table schema, automatically start to write to the Iceberg table using the new schema. This is a bit more difficult to support as the parallel writers can compete to update Iceberg tables schema. In this scenario, there is no control plane or other automation to sync up data stream schema with Iceberg table schema.

Reply:
That makes sense to me, and for now, moving the refresh back into the supplier `get()` where it was originally should be fine, as we're only using the refreshed table for the FileIO anyway right now. I'll make that change.

Reply:
I checked in that change.

Reply:
@stevenzwu: How dynamically refreshing the table (partitioning schema) between checkpoints would effect the shuffling of the writers? I think, beside the committer change, the statistics part is needed to be tweaked too to supply the 2nd type of schema evolution. Also data with dynamically changing schema without a schema registry would not be performant because of serialization issues. So I lean to supporting only the 1st version of the schema evolution ATM.

That said, I think if we agree that we are flexible about this part of the change, then we should not block this PR for this decision.

Reply:
> That said, I think if we agree that we are flexible about this part of the change, then we should not block this PR for this decision.

I was hoping to get some consensus on this topic/design (not implementation), because it affects the how we should expose this PR in the public API change `tableRefreshInteval(Duration)` vs `enableTableRefresh()`. if we always want to tie the switch at checkpoint boundary, the later boolean version seems more natural because interval is not needed at all. this seems like the only open item left. internal implementation is easy to change. public API needs to be more careful.

@pvary do you see any fundamental problems if tasks make independent timing on switching to new schema (without checkpoint boundary alignment)? I am concerned about checkpoint boundary in the future implementation of coordinator broadcast. what if some broadcast msg came before the checkpoint barrier and some come after? is it really possible for us to fully align all writer tasks perfectly? I felt it is very challenging. the alternative approach is not to rely on it. let the write tasks switch independently. we just enhance the committer to handle potentially multiple partition specs.

Reply:
My main concern before was within a single subtask, rather than across different subtasks. I felt changing the schema or spec while the file is being written could potentially lead to problems down the road, where the refresh isn't done in a deterministic way.

Even using a broadcast or token delegation framework, you could still cache and refresh in the supplier only at given times in the supplier for those. And before the latest change, no public APIs were changed.

Either way,  implementing schema evolution is going to require additional changes, so the current PR will be fine to refresh the FileIO and how the refresh occurs can be updated as needed when schema evolution is added.

Reply:
I'd actually prefer to revert back to only refreshing during the checkpoint and having a minimum refresh interval. The subtasks can get out of sync regardless, but it ensures some consistency within a subtask during a checkpoint interval.

Reply:
I checked in an alternative solution, where the table is refreshed only before creating a new task writer. This keeps refresh logic out of `IcebergStreamWriter`, but prevents the table from being updated unintentionally when getting the table reference.